lib/index.js

```diff
@@ -47,53 +47,55 @@ module.exports = function loader (css, map) {
 
   validateOptions(require('./options.json'), options, 'PostCSS Loader')
 
-  const rc = {
-    path: path.dirname(file),
-    ctx: {
-      file: {
-        extname: path.extname(file),
-        dirname: path.dirname(file),
-        basename: path.basename(file)
-      },
-      options: {}
+  Promise.resolve().then(() => {
+    if (!options.config) {
+      return parseOptions.call(this, options)
     }
-  }
 
-  if (options.config) {
-    if (options.config.path) {
-      rc.path = path.resolve(options.config.path)
+    const rc = {
+      path: path.dirname(file),
+      ctx: {
+        file: {
+          extname: path.extname(file),
+          dirname: path.dirname(file),
+          basename: path.basename(file)
+        },
+        options: {}
+      }
     }
 
-    if (options.config.ctx) {
-      rc.ctx.options = options.config.ctx
-    }
-  }
-
-  const sourceMap = options.sourceMap
-
-  Promise.resolve().then(() => {
-    const length = Object.keys(options).length
-
-    // TODO
-    // Refactor
-    if (!options.config && !sourceMap && length) {
-      return parseOptions.call(this, options)
-    } else if (options.config && !sourceMap && length > 1) {
-      return parseOptions.call(this, options)
-    } else if (!options.config && sourceMap && length > 1) {
-      return parseOptions.call(this, options)
-    } else if (options.config && sourceMap && length > 2) {
-      return parseOptions.call(this, options)
+    // Read options from options.config only when options.config is object
+    if (typeof options.config === 'object' &&
+      options.config !== null &&
+      Object.prototype.toString.call(options.config) === '[object Object]'
+    ) {
+      if (options.config.path) {
+        rc.path = path.resolve(options.config.path)
+      }
+
+      if (options.config.ctx) {
+        rc.ctx.options = options.config.ctx
+      }
     }
 
     return postcssrc(rc.ctx, rc.path, { argv: false })
   }).then((config) => {
     if (!config) config = {}
+    if (!config.options) config.options = {}
 
     if (config.file) this.addDependency(config.file)
 
+    let sourceMap = config.options.map || options.sourceMap
```

Comment:
When we load postcss config, she can return `map: undefined`, but in loader we have `sourceMap: true`, because first look on `config.options.map` then on `options.sourceMap`.